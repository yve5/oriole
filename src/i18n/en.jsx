const I18N_DICTIONARY = {
  LANG: 'en',
  LANGUAGE: 'English',
  DATE_FORMAT: 'MM/dd/yyyy',
  LOCALE: 'enUS',
  'About us': 'About us',
  'An error has been detected. Please try again.':
    'An error has been detected. Please try again.',
  Buttons: 'Buttons',
  Changelog: 'Changelog',
  Collapse: 'Collapse',
  Expand: 'Expand',
  Internationalization: 'Internationalization',
  Paging: 'Paging',
  'The request has been sent successfully.':
    'The request has been sent successfully.',
};

export default I18N_DICTIONARY;
