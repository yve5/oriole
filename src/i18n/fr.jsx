const I18N_DICTIONARY = {
  LANG: 'fr',
  LANGUAGE: 'Français',
  DATE_FORMAT: 'dd/MM/yyyy',
  LOCALE: 'fr',
  'About us': 'À propos de nous',
  'An error has been detected. Please try again.':
    'Une erreur a été détectée. Merci de recommencer.',
  Buttons: 'Boutons',
  Changelog: 'Journal des modifications',
  'Changes have been saved.': 'Les changements ont été sauvegardés.',
  'Choose a file': 'Choisir un fichier',
  Collapse: 'Réduire',
  Columns: 'Colonnes',
  'Creation date': 'Date de création',
  'Descending order': 'Ordre décroissant',
  Disabled: 'Désactivé',
  Enabled: 'Activé',
  Expand: 'Développer',
  First: 'Premier',
  'Hello World': 'Bonjour le monde',
  'Hide columns': 'Cacher les colonnes',
  Internationalization: 'Internationalisation',
  Paging: 'Pagination',
  'The request has been sent successfully.':
    'La demande a été envoyée avec succès.',
};

export default I18N_DICTIONARY;
