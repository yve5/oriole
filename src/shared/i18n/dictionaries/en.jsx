const I18N_DICTIONARY = {
  LANG: 'en',
  LANGUAGE: 'English',
  DATE_FORMAT: 'MM/dd/yyyy',
  LOCALE: 'enUS',
  '$number items per page': '$number items per page',
  '$total records': '$total records',
  Add: 'Add',
  'Ascending order': 'Ascending order',
  Back: 'Back',
  'Changes have been saved.': 'Changes have been saved.',
  'Choose a file': 'Choose a file',
  Columns: 'Columns',
  'Creation date': 'Creation date',
  'Descending order': 'Descending order',
  Disabled: 'Disabled',
  Enabled: 'Enabled',
  First: 'First',
  'Hello World': 'Hello World',
  'Hide columns': 'Hide columns',
  Last: 'Last',
  'Loading...': 'Loading...',
  Members: 'Members',
  Next: 'Next',
  No: 'Non',
  'No record': 'No record',
  'One record': 'One record',
  Previous: 'Previous',
  Reset: 'Reset',
  'This field is invalid.': 'This field is invalid.',
  Title: 'Title',
  'to fill': 'to fill',
  'Update of $update, created at $start by $guest':
    'Update of $update, created at $start by $guest',
  Yes: 'Oui',
};

export default I18N_DICTIONARY;
