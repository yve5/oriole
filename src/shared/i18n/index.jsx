import i18nState from './resources/initialState';
import I18n from './components/I18n';
import i18nReducer from './reducer';

export { I18n, i18nState, i18nReducer };
