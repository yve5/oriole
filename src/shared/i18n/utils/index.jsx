import { checkTermOrder, convertCollectionToArray } from './Term';
import { checkMissingTerms } from './Missing';
import { formatDate } from './Format';
import { tsl } from './Translate';

export {
  tsl,
  formatDate,
  checkTermOrder,
  checkMissingTerms,
  convertCollectionToArray,
};
