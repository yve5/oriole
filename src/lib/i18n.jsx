import { I18n, LangSwitch, i18nReducer } from '../shared/i18n';

export { I18n, LangSwitch, i18nReducer };
