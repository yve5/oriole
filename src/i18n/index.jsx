import fr from './fr';
import en from './en';
import { tsl } from '../shared/i18n/utils/Translate';

const i18n = {
  tsl: (term, args) => tsl(fr, term, args),
  dictionaries: [fr, en],
  lang: fr.LANG,
};

export default i18n;
