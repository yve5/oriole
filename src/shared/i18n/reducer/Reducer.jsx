import en from '../dictionaries/en';
import fr from '../dictionaries/fr';

import { tsl } from '../utils';
import {
  I18N_INIT_LANG,
  I18N_CHANGE_LANG,
  I18N_SWITCH_LANG,
  I18N_LOCAL_USER_LANG,
} from '../resources/constants';

export const initialState = {
  tsl: (term, args) => tsl(en, term, args),
  dictionaries: [en, fr],
  lang: en.LANG,
};

const reducer = (...dictionaries) => {
  let innerState = { ...initialState };

  if (dictionaries && dictionaries[0]) {
    innerState = {
      ...innerState,
      dictionaries,
      lang: dictionaries[0].LANG,
      tsl: (term, args) => tsl(dictionaries[0], term, args),
    };
  }

  return (state = innerState, action = {}) => {
    switch (action.type) {
      case I18N_INIT_LANG:
      case I18N_CHANGE_LANG: {
        const currentIndex = state.dictionaries.findIndex(
          ({ LANG }) => LANG === action.lang
        );

        return {
          ...state,
          ...action,
          tsl: (term, args) =>
            tsl(state.dictionaries[currentIndex], term, args),
        };
      }

      case I18N_SWITCH_LANG: {
        const currentIndex = state.dictionaries.findIndex(
          ({ LANG }) => LANG === state.lang
        );
        const nextIndex = (currentIndex + 1) % state.dictionaries.length;
        const nextDico = state.dictionaries[nextIndex];
        const nextLang = nextDico.LANG;

        localStorage.setItem(I18N_LOCAL_USER_LANG, nextLang);

        return {
          ...state,
          ...action,
          lang: nextLang,
          tsl: (term, args) => tsl(state.dictionaries[nextIndex], term, args),
        };
      }

      default:
        return state;
    }
  };
};

export default reducer;
