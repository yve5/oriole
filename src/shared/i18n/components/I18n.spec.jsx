import React from 'react';
import configureMockStore from 'redux-mock-store';

import { vi } from 'vitest';
import { thunk } from 'redux-thunk';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import reducer from '../reducer';
import en from '../dictionaries/en';
import fr from '../dictionaries/fr';

import { I18n } from '../../../lib/i18n';

const mockStore = configureMockStore([thunk]);

const defaultStore = mockStore({ i18n: reducer([fr, en])() });

describe('I18n', () => {
  beforeEach(() => {
    console.error = vi.fn();
  });

  it('should return a default answer', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n />
      </Provider>
    );
    const fragment = asFragment();
    expect(fragment.textContent).toBe('-');
  });

  it('should translate a string', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n>Hello World</I18n>
      </Provider>
    );
    const fragment = asFragment();
    expect(fragment.textContent).toBe('Bonjour le monde');
  });

  it('should translate an option', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n>
          <option value="42">Hello World</option>
        </I18n>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should handle an error with 2 options', () => {
    const error = () => {
      render(
        <Provider store={defaultStore}>
          <I18n>
            <option value="42">Hello World</option>
            <option value="42">Hello World</option>
          </I18n>
        </Provider>
      );
    };
    expect(error).toThrow('I18n can not translate several html tags in a row');
  });

  it('should handle an error with a wrong attribute array', () => {
    const error = () => {
      render(
        <Provider store={defaultStore}>
          <I18n attrs="Hello World">
            <img src="picture" alt="Hello World" title="Hello World" />
          </I18n>
        </Provider>
      );
    };
    expect(error).toThrow('I18n attributes are missing');
  });

  it('should handle an error with an empty attribute array', () => {
    const error = () => {
      render(
        <Provider store={defaultStore}>
          <I18n attrs={[]}>
            <img src="picture" alt="Hello World" title="Hello World" />
          </I18n>
        </Provider>
      );
    };
    expect(error).toThrow('I18n attributes are missing');
  });

  it('should translate attributes', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n attrs={['alt', 'title', 'hello']}>
          <img src="picture" alt="Hello World" title="Hello World" />
        </I18n>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should translate attributes with shortcuts', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n title>
          <img alt="Lorem ipsum..." src="picture" title="Hello World" />
        </I18n>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should translate an option with args', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n args={{ $days: '47' }}>
          <option value="42">Hello World</option>
        </I18n>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should translate a placeholder', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n placeholder>
          <input type="text" value="" placeholder="Hello World" />
        </I18n>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('I18n date parser', () => {
  const dateExample = new Date(2020, 2, 14, 13, 37, 14);

  it('should format a date', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n date={dateExample} />
      </Provider>
    );

    const fragment = asFragment();
    expect(fragment.textContent).toBe('14/03/2020');
  });

  it('should format a undefined date', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n date={undefined} />
      </Provider>
    );
    const fragment = asFragment();
    expect(fragment.textContent).toBe('-');
  });

  it('should format a date with a customized pattern', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n date={dateExample} format="PPp" />
      </Provider>
    );
    const fragment = asFragment();
    expect(fragment.textContent).toBe('14 mars 2020, 13:37');
  });

  it('should manage a weird case', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n date={dateExample}>Lorem ipsum</I18n>
      </Provider>
    );
    const fragment = asFragment();
    expect(fragment.textContent).toBe('14/03/2020');
  });

  const startExample = new Date(2017, 4, 14, 17, 51, 31);
  const updateExample = new Date(2020, 2, 14, 13, 37, 14);

  it('should format a date with expression & arguments', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n
          args={{
            $guest: 'John Doe',
            $start: { date: startExample, locale: 'fr' },
            $update: { date: updateExample, locale: 'fr' },
          }}
        >
          Update of $update, created at $start by $guest
        </I18n>
      </Provider>
    );
    const fragment = asFragment();
    expect(fragment.textContent).toBe(
      'Mise à jour le 14/03/2020, créé le 14/05/2017 par John Doe'
    );
  });

  it('should format a date with attributes & arguments', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n
          attrs={['alt', 'title', 'hello']}
          args={{
            $guest: 'John Doe',
            $start: { date: startExample, locale: 'fr' },
            $update: { date: updateExample, locale: 'fr' },
          }}
        >
          <img
            src="picture"
            alt="Hello World"
            title="Update of $update, created at $start by $guest"
          />
        </I18n>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should format a date with an input tag', () => {
    const { asFragment } = render(
      <Provider store={defaultStore}>
        <I18n
          attrs={['value']}
          args={{ $dateToUpdate: { date: startExample, locale: 'fr' } }}
        >
          <input
            disabled
            type="text"
            value="$dateToUpdate"
            className="form-control form-control-sm"
          />
        </I18n>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
