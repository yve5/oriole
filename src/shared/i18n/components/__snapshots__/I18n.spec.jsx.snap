// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`I18n > should translate a placeholder 1`] = `
<DocumentFragment>
  <input
    placeholder="Bonjour le monde"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`I18n > should translate an option 1`] = `
<DocumentFragment>
  <option
    value="42"
  >
    Bonjour le monde
  </option>
</DocumentFragment>
`;

exports[`I18n > should translate an option with args 1`] = `
<DocumentFragment>
  <option
    value="42"
  >
    Bonjour le monde
  </option>
</DocumentFragment>
`;

exports[`I18n > should translate attributes 1`] = `
<DocumentFragment>
  <img
    alt="Bonjour le monde"
    src="picture"
    title="Bonjour le monde"
  />
</DocumentFragment>
`;

exports[`I18n > should translate attributes with shortcuts 1`] = `
<DocumentFragment>
  <img
    alt="Lorem ipsum..."
    src="picture"
    title="Bonjour le monde"
  />
</DocumentFragment>
`;

exports[`I18n date parser > should format a date with an input tag 1`] = `
<DocumentFragment>
  <input
    class="form-control form-control-sm"
    disabled=""
    type="text"
    value="14/05/2017"
  />
</DocumentFragment>
`;

exports[`I18n date parser > should format a date with attributes & arguments 1`] = `
<DocumentFragment>
  <img
    alt="Bonjour le monde"
    src="picture"
    title="Mise à jour le 14/03/2020, créé le 14/05/2017 par John Doe"
  />
</DocumentFragment>
`;
